### PR TITLE
Optimized replication trigger

### PR DESCRIPTION
**Pull request by Seba-T**
Well in my defense I have empirical data showing the average replication throughput is 80% lower than the previous approach with a moderate server load. I still am waiting for the max throughput results 

- [x] This PR contains no AI generated code
- [x] `cargo fmt` has been run
- [x] `cargo clippy` has been run
- [x] `cargo test` has been run and passes
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

### CODE REVIEW

File: server/core/src/repl/mod.rs [hunk 1]
```diff
@@ -363,11 +368,60 @@ async fn repl_run_consumer(
...
 
+#[instrument(level="info", skip(idms), fields(eventid=Uuid::new_v4().to_string()))]
+async fn replication_trigger(
+    idms: &IdmServer,
+    last_replication_instant: &Instant,
+    enable_experimental_replication_trigger: bool,
+    replication_interval: Duration,
+) {
+    tokio::select! {
+        _ = async {
```
Comment (Firstyear) [inline server/core/src/repl/mod.rs:379]: This should probably be a function instead, it's really hard to parse this.

File: server/lib/src/be/idl_arc_sqlite.rs [hunk 2]
```diff
@@ -58,6 +59,7 @@ pub struct IdlArcSqlite {
     op_ts_max: CowCell<Option<Duration>>,
     allids: CowCell<IDLBitRange>,
     maxid: CowCell<u64>,
+    write_ops_since_last_repl: CowCell<u64>,
```
Comment (Firstyear) [inline server/lib/src/be/idl_arc_sqlite.rs:62]: This shouldn,t be here, it should be in the QueryServer instead.

File: server/lib/src/plugins/write_ops_counter.rs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        Self::reset_write_ops_counter(qs);
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::prelude::*;
+
+    #[qs_test(domain_level=DOMAIN_LEVEL_7)]
```
Comment (Firstyear) [inline server/lib/src/plugins/write_ops_counter.rs:103]: Don't need the domain_level set here.

File: server/core/src/repl/mod.rs [hunk 4]
```diff
@@ -363,11 +368,60 @@ async fn repl_run_consumer(
...
+            let mut one_sec_int = interval(Duration::from_secs(1));
+            // sleep for the minimum window size before we do anything
+            sleep(Duration::from_secs(MIN_REPL_TASK_POLL_INTERVAL)).await;
+            loop {
+                let write_ops_since_last_repl = read_txn.qs_read.get_write_ops_since_last_repl();
+                let time_since_last_repl = last_replication_instant.elapsed().as_secs();
+
+                let Some(write_throughput_index) = time_since_last_repl.checked_sub(MIN_REPL_TASK_POLL_INTERVAL) else {
+                    error!("Negative replication window index detected, aborting...");
+                    return;
```
Comment (Firstyear) [inline server/core/src/repl/mod.rs:391]: We should handle this error better than returning, because a return here will kill the repl task I think?

Comment (Firstyear) [reply server/core/src/repl/mod.rs:391]: Just do saturating sub instead.

File: server/core/src/repl/mod.rs [hunk 5]
```diff
@@ -363,11 +368,60 @@ async fn repl_run_consumer(
...
+                };
+
+                let (current_max_throughput, current_repl_window) = MAX_WRITE_THROUGHPUT_BY_TASK_POLL_INTERVAL[write_throughput_index as usize];
```
Comment (Firstyear) [inline server/core/src/repl/mod.rs:394]: No check on the bounds here which could cause out of bounds crash. Use .get() instead.

File: proto/src/constants.rs [hunk 6]
```diff
@@ -46,6 +46,30 @@ pub const DEFAULT_REPLICATION_ORIGIN: &str = "repl://localhost:8444";
...
+pub const MIN_REPL_TASK_POLL_INTERVAL: u64 = 3;
+
+// Estimate of max write ops that can be processed in the corresponding replication poll window
+pub const MAX_WRITE_THROUGHPUT_BY_TASK_POLL_INTERVAL: [(u64, u64); 13] = [
+    (2945, 3),
+    (4081, 4),
+    (5294, 5),
+    (6351, 6),
+    (7408, 7),
+    (8280, 8),
```
Comment (Firstyear) [inline proto/src/constants.rs:59]: Given this appears linear, it could be a function instead of a constant.

**Review by Firstyear — CHANGES_REQUESTED** [review]
It's a bit messy in it's current form - I think it can be simpler and cleaner. A lot of your tests appear redundant too?

File: proto/src/constants.rs (hunk 6 above)
Comment (yaleman) [reply proto/src/constants.rs:59]: How was this calculated, surely it's heavily dependent on system architecture etc?

File: server/core/src/repl/mod.rs (hunk 1 above)
Comment (Seba-T) [reply server/core/src/repl/mod.rs:379]: Wopsies hadn't thought of that :p. I was too enthused by finally understanding the whole tokio::select! business :D

File: server/lib/src/plugins/write_ops_counter.rs (hunk 3 above)
Comment (Seba-T) [reply server/lib/src/plugins/write_ops_counter.rs:103]: I must admit I blindly copied & pasted from the other tests, not having a clue how the macro works :D

File: server/core/src/repl/mod.rs (hunk 4 above)
Comment (Seba-T) [reply server/core/src/repl/mod.rs:391]: > We should handle this error better than returning, because a return here will kill the repl task I think?

I mean I haven't tested it but I assumed it would just immediately return the future? I'd expect the repl loop to resume normally. Actaully if returning kills the entire repl task, then also this [return](https://github.com/Seba-T/kanidm/blob/53a866bdc0a8b6d47c5aa7c8a1f7a3239df3b12e/server/core/src/repl/mod.rs#L414) should kill it, but I'm pretty sure it doesn't since I've tested it.

File: server/core/src/repl/mod.rs (hunk 5 above)
Comment (Seba-T) [reply server/core/src/repl/mod.rs:394]: wopsies

File: proto/src/constants.rs (hunk 6 above)
Comment (Seba-T) [reply proto/src/constants.rs:59]: > How was this calculated, surely it's heavily dependent on system architecture etc?

The numbers come from our benchmarks, but you're right in saying they are ... not consistent across different machines. But I have another secret plan to fix that ;-). For now I think it makes sense to just assume they're correct and polish everything else.

Comment (Seba-T) [reply proto/src/constants.rs:59]: > Given this appears linear, it could be a function instead of a constant.

Sure, but I think we first need the big chonky benchmarks we were discussing yesterday before interpolating the points. Also we wouldn't need to compute weird indexes as we're currently doing, it would be much more elegant!

**Comment by Seba-T** [comment]
> It's a bit messy in it's current form - I think it can be simpler and cleaner. A lot of your tests appear redundant too?

I think they all can be coalesced into one per section, i.e one repl test, one plugin test, I kept them separate because I thought it was easier to understand, but maybe I should merge them?

File: server/lib/src/be/idl_arc_sqlite.rs (hunk 2 above)
Comment (Seba-T) [reply server/lib/src/be/idl_arc_sqlite.rs]: sure!

File: proto/src/constants.rs [hunk 7]
```diff
@@ -43,9 +43,33 @@ pub const DEFAULT_LDAP_LOCALHOST: &str = "localhost:636";
...
+    (11919, 11),
+    (12395, 12),
+    (12770, 13),
+    (14293, 14),
+    (15893, 15),
+];
+
+// This represents the percentage of the current MAX_WRITE_THROUGHPUT (determined by the current replication interval)
+// above which we will not trigger replication as we are too close to the maximum throughput and we risk to lower the server performance
+pub const REPLICATION_TRIGGER_SAFETY_THRESHOLD: f64 = 0.8;
```
Comment (yaleman) [inline proto/src/constants.rs:71]: So we're not going over 0.8% throughput? I'm guessing you mean the *ratio* of `MAX_WRITE_THROUGHPUT` rather than percentage?

Comment (Seba-T) [reply proto/src/constants.rs:71]: Yess it's definitely not 0.8% :D. It should say ratio instead.

**Comment by Firstyear** [comment]
So far I'm skeptical of the possible gains here because as @yaleman said these numbers are specific to the system you benchmarked on, and your benchmarks so far aren't consistent or widely sampled enough to make a firm assessment of an optimal value. I think we delay this until we have more conclusive evidence to support it's inclusion.

File: proto/src/constants.rs (hunk 6 above)
Comment (yaleman) [reply proto/src/constants.rs:59]: >  For now I think it makes sense to just assume they're correct and polish everything else.

I'll have to disagree on that one, for now.

Comment (Firstyear) [reply proto/src/constants.rs:59]: As do I.

**Comment by Firstyear** [comment]
Closing as I think this isn't needed.
